File: 2000/LC 1690.cpp
class Solution {
public:
    int stoneGameVII(vector<int>& a) {
        int n = a.size();
        vector<int> sum(n + 1, 0);
        vector<int> dp(n);
        for (int i = 0; i < n; ++i) {
            sum[i+1] = sum[i] + a[i];
        }
        for (int i = n - 2; i >= 0; --i) {
            for (int j = i + 1; j < n; ++j) {
                dp[j] = max(sum[j] - sum[i] - dp[j-1], sum[j+1] - sum[i+1] - dp[j]);
            }
        }
        return dp[n-1];
    }
};

File: 2000/LC 1690.go
func stoneGameVII(stones []int) int {
    n := len(stones)
    pre := make([]int, n + 1)
    for i, x := range stones {
        pre[i+1] = pre[i] + x
    }

    f := make([][]int, n)
    for i := n - 1; i >= 0; i-- {
        f[i] = make([]int, n)
        for j := i + 1; j < n; j++ {
            f[i][j] = max(pre[j+1] - pre[i+1] - f[i+1][j], pre[j] - pre[i] - f[i][j-1])
        }
    }

    return f[0][n-1]
}